Implement Python bindings `pymetacg` (#32)
tudasc/MetaCG: pymetacg/README.md

# pymetacg

`pymetacg` are MetaCG's Python bindings that support parsing MetaCG files from Python projects.
Currently, manipulating and writing call graphs is *not* supported.

## Dependencies
 - [nanobind](https://github.com/wjakob/nanobind)
 - [pytest](https://pypi.org/project/pytest/) (for tests only)

>  - [pytext-cmake](https://pypi.org/project/pytest/) (for tests only)

Reviewer: Typo and link wrong

Author: Fixed

## Building
`pymetacg` can be built as an optional feature during the MetaCG build, producing a dynamic library (e.g., `pymetacg.cpython-313-x86_64-linux-gnu.so`) that can then be loaded from Python.
Pass `-DMETACG_BUILD_PYMETACG=ON` to CMake to enable `pymetacg`.

```bash
$> cmake -S . -B build -DCMAKE_INSTALL_PREFIX=/where/to/install -DMETACG_BUILD_PYMETACG=ON
$> cmake --build build --parallel
$> cmake --install build
```

By default, nanobind is downloaded automatically during configruation.
To use an already installed nanobind, use `-DMETACG_USE_EXTERNAL_NANOBIND=ON` and specify the installation path using `-Dnanobind_ROOT=<your nanobind installation>`, if necessary.

The CMake setup of `pymetacg` uses CMake's [FindPython](https://cmake.org/cmake/help/latest/module/FindPython.html) to find the Python installation for which `pymetacg` will be built.
To specify which Python installation to use, use `-DPython_ROOT_DIR=<python installation>`.

> By default, the resulting dynamic library will be installed in `<installation prefix>/lib/<python version>/site-packages`.

Reviewer: What is the `<installation prefix>` mentioned here?

Reviewer: Is this the MetaCG install prefix? If so, maybe we should be explicit.

Author: Done

To override the installation path, use `-DCMAKE_INSTALL_PYTHON_LIBDIR`.

> Finally, make sure that the installation directory in found by Python, e.g., by using the `PYTHONPATH` environment variable.

Reviewer: ```suggestion
Finally, make sure that the installation directory is found by Python, e.g., by using the `PYTHONPATH` environment variable.
```

Test whether `pymetacg` was built successfully:
```bash
$> python -c "import pymetacg; print(pymetacg.info)"
```

## Usage
```python
import pymetacg

# print MetaCG version information
print(pymetacg.info)

# read MetaCG file (.mcg)
cg = pymetacg.Callgraph.from_file("/path/to/callgraph.mcg")

# check whether node exists in graph
main_exists = "main" in cg

# access nodes in graph
main = cg["main"]

> foo = cg["main"]

Reviewer: ```suggestion
foo = cg["foo"]
```

# iterate over node callees and collect callee function names into list
callees_names = [callee.function_name for callee in main.callees]

# iterate over node callers and collect caller function names into list
callers_names = [caller.function_name for caller in foo.callers]

# check whether function calls itself
foo_is_recursive = foo in foo.callees

# iterate over nodes in graph and print function names
for node in cg:
    print(node.function_name)

# check if node has some meta data
main_has_statement_metadata = "numStatements" in main.meta_data

# print node meta data

> print(main.meta_data["numStatements"].data)

Reviewer: Is this always `.data`?

Author: Yes, unfortunately. I've spent a considerable amount of time trying to expose the JSON-data of meta data objects directly (so without going through `.data`), but couldn't quite figure out how to do that with nanobind. I agree that being able to omit `.data` would be nicer, so if anybody knows how to do that, I would be very interested.

Reviewer: Ok, but it is the same for all `MetaData`? That was my main question.
I think mentioning that somehow would be good for the uneducated like me.

Author: Is it clearer now?

Reviewer: Yes, thank you

```

## Tests

> `pymetacg` comes with a PyTest-based test suite.

Reviewer: ```suggestion
`pymetacg` comes with a PyTest-based test suite, which can be enabled via the CMake option `-DMETACG_BUILD_PYMETACG_TESTS=ON`.
```
Text is really just a suggestion. Not sure what's best.

> Use `-DMETACG_BUILD_PYMETACG_TESTS=ON` and install `pytest` and `pytest-cmake` need to be installed from PyPI to enable building to the tests:

Reviewer: ```suggestion
The test suite requires `pytest` and `pytest-cmake` to be installed from PyPI.
```
